File: loglite/migrations.py
from typing import List, Dict, Any, TypedDict
from loguru import logger

from .database import Database
from .types import Migration


class MigrationManager:
    def __init__(self, db: Database, migrations_config: List[Dict[str, Any]]):
        self.db = db
        self.migrations_config = migrations_config

    async def apply_pending_migrations(self) -> bool:
        """Apply all pending migrations"""
        applied_versions = await self.db.get_applied_versions()

        # Sort migrations by version
        sorted_migrations = sorted(self.migrations_config, key=lambda m: m["version"])

        success = True
        for migration in sorted_migrations:
            version = migration["version"]
            if version not in applied_versions:
                logger.info(f"Applying migration version {version}")
                statements = migration.get("rollout", [])
                if statements:
                    if not await self.db.apply_migration(version, statements):
                        success = False
                        break

        return success

    async def rollback_migration(self, version: int) -> bool:
        """Rollback a specific migration version"""
        applied_versions = await self.db.get_applied_versions()

        if version not in applied_versions:
            logger.warning(
                f"Migration version {version} not applied, nothing to rollback"
            )
            return False

        for migration in self.migrations_config:
            if migration["version"] == version:
                statements = migration.get("rollback", [])
                if statements:
                    return await self.db.rollback_migration(version, statements)
                else:
                    logger.warning(f"No rollback statements for version {version}")
                    return False

        logger.warning(f"Migration version {version} not found in configuration")
        return False

File: cpp/src/migrations.cpp
#include "migrations.hpp"
#include "writer_database.hpp"
#include "log.hpp"
#include "utils.hpp"

#include <algorithm>
#include <fmt/format.h>
#include <iostream>
#include <ranges>
#include <stdexcept>

namespace loglite {

MigrationManager::MigrationManager(WriterDatabase& db, std::span<const Migration> migrations)
    : db_(db), migrations_(migrations.begin(), migrations.end()) {
    std::ranges::sort(migrations_, {}, &Migration::version);
}

bool MigrationManager::ApplyPendingMigrations(int start_version) {
    auto applied = db_.GetAppliedVersions();

    for (const auto& mg : migrations_) {
        if (mg.version <= start_version) continue;
        if (range_contains(applied, mg.version)) continue;

        bool ok = db_.ApplyMigration(mg.version, mg.rollout);
        // Mirror Python: apply ONE migration per call and return.
        return ok;
    }
    return false;
}

bool MigrationManager::RollbackMigration(int version, bool force) {
    auto it = std::ranges::find_if(migrations_,
                                   [version](const Migration& m) { return m.version == version; });
    if (it == migrations_.end())
        throw std::runtime_error(fmt::format("Migration v{} not found in config", version));

    if (!force) {
        std::cout << fmt::format("Roll back migration v{}? [y/N] ", version);
        std::string ans;
        std::getline(std::cin, ans);
        if (ans != "y" && ans != "Y") {
            log::info("Rollback cancelled.");
            return false;
        }
    }

    return db_.RollbackMigration(version, it->rollback);
}

}  // namespace loglite
